CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI

veronikaslc opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/CARDS-1844)

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -93,7 +105,7 @@ function ToUConfiguration() {
   }
 
-  return (
+  return (<>
```
Why did the React fragment need to be added here?

veronikaslc committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI

Removed useless tags

sdumitriu committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -108,6 +120,48 @@ function ToUConfiguration() {
+              <MarkdownText
+                value={text}
+                height={450}
+                onChange={value => { event.preventDefault(); setText(value); }}
```
Why is `preventDefault`called here?

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -108,6 +120,48 @@ function ToUConfiguration() {
+                type="text"
+                label="Title"
+                value={title}
+                onChange={(event) => { setTitle(event.target.value); }}
```
Changing the "Enabled" checkbox resets the button from "Saved" to "Save", but changing the title, version, or text doesn't.

sdumitriu committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
bugfix: changing the ToU settings after saving won't change the button back to "Save"

sdumitriu reviewed (changes requested):
Minor bugs:
- deleting all the text in the ToU text will shrink the editor a lot; it only grows when there's plenty of text in it
- bold text in the preview doesn't look bold at all

veronikaslc commented:
`bold text in the preview doesn't look bold at all`
I guess it's a browser-specific. Looks ok in the Chrome.

marta- committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
* Reorder form entries more logically ("enabled" at the end)
* Restyle labels for a cleaner look
* Bug fix: prevent md editor from shrinking if text is empty

marta- committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
Update the admin entry description

marta- commented:
> * bold text in the preview doesn't look bold at all

Looks very bold on my device. @sdumitriu please post a screenshot for comparison.

![image](https://user-images.githubusercontent.com/651980/177439341-cbbb075e-ff1e-446a-86ca-11841495230e.png)

marta- committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
Rename the boolean property that states whether patient users must accept the terms of use before answering any surveys

marta- commented:
> > I think that `enabled` at the top makes more sense. It would be even better if de-selecting it would hide all the other settings, since they are useless if the whole thing is not enabled.
> 
> They are not useless. There is a Terms of Use and Privacy Policy link at the bottom of every patient screens that reveals the terms. The "enabled" property (which should probably be renamed for clarity) is not about whether Terms of Use exist. It is only about whether the patient user is forced to accept them before answering surveys

`enabled` renamed to `enforceAccept` in 18ca79d.

marta- commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -92,17 +109,52 @@ function ToUConfiguration() {
-                    onChange={(event) => { event.preventDefault(); setIsSaved(false); setEnabled(event.target.checked); }}
-                    name="enabled"
+                    checked={enforceAccept}
+                    onChange={(event) => { event.preventDefault(); setIsSaved(false); setEnforceAccept(event.target.checked); }}
```
@veronikaslc what was the purpose of `event.preventDefault()` here?

sashaandjic commented:
Firefox vs 
![image](https://user-images.githubusercontent.com/23243244/177602544-94ee2f57-d04c-416b-b9ad-728afc9cf8bb.png)

marta- commented:
> Firefox vs Chrome - bold fonts in FF is not prominent as in Chrome
 
I see, thank you @sashaandjic ! Please report this separately, we'll have to tweak the MDEditor style.

marta- committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
Removed unnecessary code

marta- commented:
> Tested - all looks and works well One very small thing that should not hold this PR is that when UoT is empty and flag `Patients must accept the Terms of Use before using the portal` is set to `true` the patient who is signing in for the first time will see this:
> 
> [image](https://user-images.githubusercontent.com/23243244/177641261-bee3a303-05b4-49f3-83ed-33ebf6879002.png)
> 
> Nothing is broken here. Everything works as per design.

I think this is ok since it's an admin setting and if the admin decides the terms should be empty the consequences are really their fault.

sdumitriu commented on proms-resources/frontend/src/main/resources/SLING-INF/content/Extensions/AdminDashboard/ToUConfiguration.json:
```diff
@@ -4,6 +4,6 @@
   "cards:targetURL": "/content.html/admin/ToUConfiguration",
   "cards:icon": "asset:proms-homepage.ToUConfigurationIcon.js",
-  "cards:hint": "Configuration for whether patient users need to accept the Terms of Use before using the Patient Portal",
+  "cards:hint": "Configure the Patient Portal Terms of Use contents and acceptance requirements",
```
```suggestion
  "cards:hint": "Configure the Patient Portal Terms of Use",
```

marta- replied:
ok.

marta- replied:
Actually no, I disagree. It looks cleaner but it doesn't bring any additional information compared to the section label. `Patient Portal Terms of Use` / `Configure the Patient Portal Terms of Use` - of course that's what we're configuring. If we say _what_ about the Terms of Use we can configure, that's when the description actually becomes useful. Feel free to rephrase as long as it says a bit more than the section label.

sdumitriu replied:
Oh, that's the hint, not the label. Nevermind.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -92,17 +109,52 @@ function ToUConfiguration() {
-        {error && <Alert severity="error">{error}</Alert>}
+    <AdminScreen title="Patient Portal Terms of Use">
+      { error && <Alert severity="error">{error}</Alert> }
+      { typeof(text) == 'undefined' ? <CircularProgress /> :
```
What happens if you delete the whole text, save, and reload the config page?

marta- replied:
The config page is displayed with an empty editor. Because the text is set to `""` if the property is missing in the json (see line 64).

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/proms/ToUConfiguration.jsx:
```diff
@@ -61,7 +75,10 @@ function ToUConfiguration() {
     // We need to do this because sling does not accept JSON, need url encoded data
     let formData = new URLSearchParams();
-    formData.append('enabled', enabled);
+    formData.append('enforceAccept', enforceAccept);
```
Will this send `undefined` to the server if any is missing?

marta- replied:
Introduced default values for all configurable ToU fields, including `false` for this one, in e731d20.

sashaandjic commented:
@marta-  I would live it as is.

sdumitriu commented on proms-resources/clinical-data/src/main/resources/SLING-INF/content/Homepage/Proms/TermsOfUse.xml:
```diff
@@ -30,7 +30,7 @@
     </property>
     <property>
-        <name>enabled</name>
+        <name>enforceAccept</name>
```
This sounds weird. enforceAcceptance? acceptanceRequired?

marta- replied:
Renamed `enforceAccept` to `acceptanceRequired` in e731d20.

marta- committed:
CARDS-1844: As an administrator, I can update the Terms of Use from the admin UI
Addressing review comments:
* Renamed `enforceAccept` to `acceptanceRequired`
* Introduced default values for all configurable ToU fields

sdumitriu commented:
@marta- Bold in Firefox:

Bold text from Markdown is just rendered as a `<strong>` element, with no styles other than what the browser has as a default.

In Firefox, this is `font-weight: bolder`, which means "increase the font weight a bit from what was inherited". This is correct.

In Chrome, it's `font-weight: bold`, which has a fixed value.

When it is displayed normally in a page (not in the editor), it is usually wrapped with a `Typography`, which does define the font weight as 400. In the editor, it is not surrounded by `Typography`, so the base font-weight is inherited all the way from `<body>`, which for an unknown reason is set to `300` instead of the default `400`. `bolder` then turns the font weight for `<strong>` to `400`, so regular text has weight 300, bold text has weight 400 in Firefox. Depending on the system, 400 may be different from 300, but not on my system, so there is no difference at all.

The question is, why is the root font weight set to 300? Is it OK if we remove this custom style from [the body element](https://github.com/data-team-uhn/cards/blob/a21f74614803499e9584c6cd62ab9415084f25b0/modules/commons/src/main/resources/SLING-INF/content/libs/cards/Resource/header.html#L56)?

marta- commented:
> @marta- Bold in Firefox:

@sdumitriu This discussion belongs on [the dedicated jira task CARDS-1852](https://phenotips.atlassian.net/browse/CARDS-1852).
